src/validation.cpp

```diff
@@ -823,43 +831,67 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     }
 
     m_rbf = !ws.m_conflicts.empty();
-    if (m_rbf) {
-        CFeeRate newFeeRate(ws.m_modified_fees, ws.m_vsize);
-        // It's possible that the replacement pays more fees than its direct conflicts but not more
-        // than all conflicts (i.e. the direct conflicts have high-fee descendants). However, if the
-        // replacement doesn't pay more fees than its direct conflicts, then we can be sure it's not
-        // more economically rational to mine. Before we go digging through the mempool for all
-        // transactions that would need to be removed (direct conflicts and all descendants), check
-        // that the replacement transaction pays more than its direct conflicts.
-        if (const auto err_string{PaysMoreThanConflicts(ws.m_iters_conflicting, newFeeRate, hash)}) {
-            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "insufficient fee", *err_string);
-        }
+    return true;
+}
 
-        // Calculate all conflicting entries and enforce BIP125 Rule #5.
-        if (const auto err_string{GetEntriesForConflicts(tx, m_pool, ws.m_iters_conflicting, ws.m_all_conflicting)}) {
-            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY,
-                                 "too many potential replacements", *err_string);
-        }
-        // Enforce BIP125 Rule #2.
-        if (const auto err_string{HasNoNewUnconfirmed(tx, m_pool, ws.m_iters_conflicting)}) {
-            return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY,
-                                 "replacement-adds-unconfirmed", *err_string);
-        }
+bool MemPoolAccept::ReplacementChecks(Workspace& ws)
+{
+    AssertLockHeld(cs_main);
```

Comment:
Would it make sense to add `Assume/Assert(ws.m_rbf);`?

Reply:
Sure, there's no harm in adding one. If safety is a concern, note that this is a no-op if `m_iters_conflicting` is empty:`PaysMoreThanConflicts`, `GetEntriesForConflicts`, `HasNoNewUnconfirmed`, and `PaysForRBF` should all simply return `std::nullopt` if they're called with empty or 0 inputs. So if we accidentally called this function with `m_rbf` false or no conflicts, it would be some small wasted effort, but we should be okay.